File: server/aws.py
from datetime import datetime

import boto3


def upload_to_s3(key_material, key_name):
    """Upload key to S3 and return the download URL"""
    s3 = boto3.client('s3')
    bucket_name = 'hng12-devbotops'
    key_path = f'keys/{key_name}.pem'
    
    try:
        s3.put_object(
            Bucket=bucket_name,
            Key=key_path,
            Body=key_material,
            ContentType='text/plain'
        )
        
        # Generate presigned URL that expires in 1 hour
        url = s3.generate_presigned_url(
            'get_object',
            Params={'Bucket': bucket_name, 'Key': key_path},
            ExpiresIn=3600
        )
        return url
    except Exception as e:
        raise Exception(f"Failed to upload key to S3: {str(e)}")


def setup_aws_instance(
    instance_type="t3.micro",
    security_group_id="sg-09bd2a53c2f724517",
):
    """
    Sets up an AWS EC2 instance and returns its details.
    Returns a JSON with instance information and credentials.
    """
    ec2 = boto3.client("ec2", region_name="us-east-2")

    timestamp = datetime.now().strftime("%m%d-%H%M%S%f")[:-3]
    key_name = f"key-{timestamp}"

    try:
        key_pair = ec2.create_key_pair(KeyName=key_name)
        
        # Upload key to S3 instead of saving locally
        key_url = upload_to_s3(key_pair["KeyMaterial"], key_name)

        ami_id = get_ubuntu_ami()
        security_group_id = ensure_security_group()
        instance = ec2.run_instances(
            ImageId=ami_id,
            InstanceType=instance_type,
            SecurityGroupIds=[security_group_id],
            KeyName=key_name,
            MinCount=1,
            MaxCount=1,
        )

        instance_id = instance["Instances"][0]["InstanceId"]

        waiter = ec2.get_waiter("instance_running")
        waiter.wait(InstanceIds=[instance_id])

        response = ec2.describe_instances(InstanceIds=[instance_id])
        public_ip = response["Reservations"][0]["Instances"][0]["PublicIpAddress"]

        return {
            "instance_id": instance_id,
            "key_id": key_name,
            "username": "ubuntu",
            "ip_address": public_ip,
            "key_url": key_url,
        }

    except Exception as e:
        try:
            ec2.delete_key_pair(KeyName=key_name)
            # Delete from S3 if exists
            s3 = boto3.client('s3')
            s3.delete_object(Bucket='hng12', Key=f'keys/{key_name}.pem')
        except:
            pass
        raise Exception(f"Failed to setup AWS instance: {str(e)}")


def get_ubuntu_ami():
    """Get latest Ubuntu 22.04 AMI ID"""
    ec2 = boto3.client("ec2", region_name="us-east-2")
    response = ec2.describe_images(
        Filters=[
            {
                "Name": "name",
                "Values": ["ubuntu/images/hvm-ssd/ubuntu-jammy-22.04-amd64-server-*"],
            },
            {"Name": "state", "Values": ["available"]},
            {"Name": "architecture", "Values": ["x86_64"]},
        ],
        Owners=["099720109477"],
    )
    return sorted(
        response["Images"], key=lambda x: x["CreationDate"], reverse=True
    )[0]["ImageId"]


def ensure_security_group(group_name="default-sg"):
    """Create security group if it doesn't exist"""
    ec2 = boto3.client("ec2", region_name="us-east-2")

    vpc_response = ec2.describe_vpcs(
        Filters=[{"Name": "isDefault", "Values": ["true"]}]
    )
    vpc_id = vpc_response["Vpcs"][0]["VpcId"]

    try:
        response = ec2.describe_security_groups(
            Filters=[{"Name": "group-name", "Values": [group_name]}]
        )
        return response["SecurityGroups"][0]["GroupId"]
    except:
        response = ec2.create_security_group(
            GroupName=group_name,
            Description="Security group for EC2 instances",
            VpcId=vpc_id,
        )
        security_group_id = response["GroupId"]

        ec2.authorize_security_group_ingress(
            GroupId=security_group_id,
            IpPermissions=[
                {
                    "IpProtocol": "tcp",
                    "FromPort": 22,
                    "ToPort": 22,
                    "IpRanges": [{"CidrIp": "0.0.0.0/0"}],
                }
            ],
        )
        return security_group_id

def destroy_all_instances():
    """Destroys all EC2 instances and keys in us-east-1 and us-east-2, and cleans up S3 keys"""
    regions = ['us-east-1', 'us-east-2']
    
    for region in regions:
        ec2 = boto3.client('ec2', region_name=region)
        
        # Get all instances
        instances = ec2.describe_instances()
        
        # Collect all instance IDs
        instance_ids = []
        for reservation in instances['Reservations']:
            for instance in reservation['Instances']:
                if instance['State']['Name'] != 'terminated':
                    instance_ids.append(instance['InstanceId'])
        
        # Terminate instances if any exist
        if instance_ids:
            try:
                ec2.terminate_instances(InstanceIds=instance_ids)
                print(f"Terminated {len(instance_ids)} instances in {region}")
            except Exception as e:
                print(f"Error terminating instances in {region}: {str(e)}")
        
        # Delete all key pairs
        try:
            keys = ec2.describe_key_pairs()
            for key in keys['KeyPairs']:
                ec2.delete_key_pair(KeyName=key['KeyName'])
                print(f"Deleted key pair {key['KeyName']} in {region}")
        except Exception as e:
            print(f"Error deleting keys in {region}: {str(e)}")
    
    # Clean up S3 keys
    try:
        s3 = boto3.client('s3')
        response = s3.list_objects_v2(Bucket='hng12-devbotops', Prefix='keys/')
        if 'Contents' in response:
            for obj in response['Contents']:
                s3.delete_object(Bucket='hng12-devbotops', Key=obj['Key'])
            print("Cleaned up keys from S3")
    except Exception as e:
        print(f"Error cleaning up S3 keys: {str(e)}")


if __name__ == "__main__":
    destroy_all_instances()

File: src/services/aws.ts
import AWS from 'aws-sdk';
import { logger, Config } from '@/config';
import { ExternalServiceError, type AwsInstanceResult } from '@/types';

interface AwsInstanceConfig {
  readonly instanceType?: string;
  readonly securityGroupId?: string;
  readonly region?: string;
}

export class AwsService {
  private readonly ec2: AWS.EC2;
  private readonly s3: AWS.S3;

  constructor(private readonly config: AwsInstanceConfig = {}) {
    const region = config.region ?? Config.aws.region;

    AWS.config.update({ region });

    this.ec2 = new AWS.EC2({ region });
    this.s3 = new AWS.S3({ region });
  }

  /**
   * Upload SSH key to S3 and return presigned URL
   */
  private async uploadKeyToS3(keyMaterial: string, keyName: string): Promise<string> {
    try {
      const keyPath = `keys/${keyName}.pem`;

      await this.s3
        .putObject({
          Bucket: Config.aws.s3Bucket,
          Key: keyPath,
          Body: keyMaterial,
          ContentType: 'text/plain',
        })
        .promise();

      const url = this.s3.getSignedUrl('getObject', {
        Bucket: Config.aws.s3Bucket,
        Key: keyPath,
        Expires: 3600, // 1 hour
      });

      return url;
    } catch (error) {
      throw new ExternalServiceError(
        `Failed to upload key to S3: ${error instanceof Error ? error.message : String(error)}`,
        'AWS-S3'
      );
    }
  }

  /**
   * Get the latest Ubuntu 22.04 AMI ID
   */
  private async getUbuntuAmi(): Promise<string> {
    try {
      const result = await this.ec2
        .describeImages({
          Filters: [
            {
              Name: 'name',
              Values: ['ubuntu/images/hvm-ssd/ubuntu-jammy-22.04-amd64-server-*'],
            },
            { Name: 'state', Values: ['available'] },
            { Name: 'architecture', Values: ['x86_64'] },
          ],
          Owners: ['099720109477'], // Canonical
        })
        .promise();

      const images = result.Images ?? [];
      if (images.length === 0) {
        throw new Error('No Ubuntu AMIs found');
      }

      const latestImage = images.sort(
        (a, b) => new Date(b.CreationDate ?? 0).getTime() - new Date(a.CreationDate ?? 0).getTime()
      )[0];

      if (!latestImage?.ImageId) {
        throw new Error('No valid Ubuntu AMI found');
      }

      return latestImage.ImageId;
    } catch (error) {
      throw new ExternalServiceError(
        `Failed to get Ubuntu AMI: ${error instanceof Error ? error.message : String(error)}`,
        'AWS-EC2'
      );
    }
  }

  /**
   * Ensure security group exists or create it
   */
  private async ensureSecurityGroup(groupName: string = 'default-sg'): Promise<string> {
    try {
      // Get default VPC
      const vpcResult = await this.ec2
        .describeVpcs({
          Filters: [{ Name: 'isDefault', Values: ['true'] }],
        })
        .promise();

      const defaultVpc = vpcResult.Vpcs?.[0];
      if (!defaultVpc?.VpcId) {
        throw new Error('No default VPC found');
      }

      // Check if security group exists
      try {
        const sgResult = await this.ec2
          .describeSecurityGroups({
            Filters: [{ Name: 'group-name', Values: [groupName] }],
          })
          .promise();

        const existingSg = sgResult.SecurityGroups?.[0];
        if (existingSg?.GroupId) {
          return existingSg.GroupId;
        }
      } catch {
        // Security group doesn't exist, will create it
      }

      // Create security group
      const createResult = await this.ec2
        .createSecurityGroup({
          GroupName: groupName,
          Description: 'Security group for EC2 instances',
          VpcId: defaultVpc.VpcId,
        })
        .promise();

      const securityGroupId = createResult.GroupId;
      if (!securityGroupId) {
        throw new Error('Failed to create security group');
      }

      // Add SSH access rule
      await this.ec2
        .authorizeSecurityGroupIngress({
          GroupId: securityGroupId,
          IpPermissions: [
            {
              IpProtocol: 'tcp',
              FromPort: 22,
              ToPort: 22,
              IpRanges: [{ CidrIp: '0.0.0.0/0' }],
            },
          ],
        })
        .promise();

      return securityGroupId;
    } catch (error) {
      throw new ExternalServiceError(
        `Failed to ensure security group: ${error instanceof Error ? error.message : String(error)}`,
        'AWS-EC2'
      );
    }
  }

  /**
   * Setup AWS EC2 instance with SSH key
   */
  async setupInstance(): Promise<AwsInstanceResult> {
    const timestamp = new Date().toISOString().replace(/[:.]/g, '-').slice(0, -5);
    const keyName = `key-${timestamp}`;

    try {
      logger.info('Creating key pair...');
      const keyPair = await this.ec2.createKeyPair({ KeyName: keyName }).promise();

      if (!keyPair.KeyMaterial) {
        throw new Error('No key material returned from AWS');
      }

      logger.info('Uploading key to S3...');
      const keyUrl = await this.uploadKeyToS3(keyPair.KeyMaterial, keyName);

      logger.info('Getting Ubuntu AMI...');
      const amiId = await this.getUbuntuAmi();

      logger.info('Ensuring security group...');
      const securityGroupId = this.config.securityGroupId ?? (await this.ensureSecurityGroup());

      logger.info('Launching EC2 instance...');
      const instanceResult = await this.ec2
        .runInstances({
          ImageId: amiId,
          InstanceType: this.config.instanceType ?? Config.aws.defaultInstanceType,
          SecurityGroupIds: [securityGroupId],
          KeyName: keyName,
          MinCount: 1,
          MaxCount: 1,
        })
        .promise();

      const instance = instanceResult.Instances?.[0];
      if (!instance?.InstanceId) {
        throw new Error('No instance returned from AWS');
      }

      logger.info(`Waiting for instance ${instance.InstanceId} to be running...`);
      await this.ec2
        .waitFor('instanceRunning', {
          InstanceIds: [instance.InstanceId],
        })
        .promise();

      logger.info('Getting instance public IP...');
      const describeResult = await this.ec2
        .describeInstances({
          InstanceIds: [instance.InstanceId],
        })
        .promise();

      const runningInstance = describeResult.Reservations?.[0]?.Instances?.[0];
      if (!runningInstance?.PublicIpAddress) {
        throw new Error('No public IP address found for instance');
      }

      const result: AwsInstanceResult = {
        instanceId: instance.InstanceId,
        keyId: keyName,
        username: 'ubuntu',
        ipAddress: runningInstance.PublicIpAddress,
        keyUrl,
      };

      logger.info('EC2 instance setup completed successfully:', result);
      return result;
    } catch (error) {
      // Cleanup on failure
      try {
        await this.ec2.deleteKeyPair({ KeyName: keyName }).promise();
        await this.s3
          .deleteObject({
            Bucket: Config.aws.s3Bucket,
            Key: `keys/${keyName}.pem`,
          })
          .promise();
      } catch (cleanupError) {
        logger.error('Error during cleanup:', cleanupError);
      }

      throw new ExternalServiceError(
        `Failed to setup AWS instance: ${error instanceof Error ? error.message : String(error)}`,
        'AWS-EC2'
      );
    }
  }

  /**
   * Destroy all instances and cleanup resources
   */
  async destroyAllInstances(regions: string[] = ['us-east-1', 'us-east-2']): Promise<void> {
    const cleanupPromises = regions.map(async region => {
      const ec2 = new AWS.EC2({ region });

      try {
        // Get all instances
        const instancesResult = await ec2.describeInstances().promise();
        const instanceIds: string[] = [];

        for (const reservation of instancesResult.Reservations ?? []) {
          for (const instance of reservation.Instances ?? []) {
            if (instance.State?.Name !== 'terminated' && instance.InstanceId) {
              instanceIds.push(instance.InstanceId);
            }
          }
        }

        // Terminate instances
        if (instanceIds.length > 0) {
          await ec2.terminateInstances({ InstanceIds: instanceIds }).promise();
          logger.info(`Terminated ${instanceIds.length} instances in ${region}`);
        }

        // Delete all key pairs
        const keysResult = await ec2.describeKeyPairs().promise();
        for (const key of keysResult.KeyPairs ?? []) {
          if (key.KeyName) {
            await ec2.deleteKeyPair({ KeyName: key.KeyName }).promise();
            logger.info(`Deleted key pair ${key.KeyName} in ${region}`);
          }
        }
      } catch (error) {
        logger.error(`Error cleaning up resources in ${region}:`, error);
      }
    });

    // Cleanup S3 keys
    try {
      const s3Objects = await this.s3
        .listObjectsV2({
          Bucket: Config.aws.s3Bucket,
          Prefix: 'keys/',
        })
        .promise();

      if (s3Objects.Contents) {
        const deleteObjects = s3Objects.Contents.filter(obj => obj.Key).map(obj => ({
          Key: obj.Key!,
        }));

        if (deleteObjects.length > 0) {
          await this.s3
            .deleteObjects({
              Bucket: Config.aws.s3Bucket,
              Delete: { Objects: deleteObjects },
            })
            .promise();
          logger.info('Cleaned up keys from S3');
        }
      }
    } catch (error) {
      logger.error('Error cleaning up S3 keys:', error);
    }

    await Promise.all(cleanupPromises);
  }
}
